fdbclient/FileBackupAgent.actor.cpp

```diff
@@ -2997,18 +2997,13 @@ struct BackupSnapshotDispatchTask : BackupTaskFuncBase {
 		// In this context "all" refers to all of the shards relevant for this particular backup
 		state int countAllShards = countShardsDone + countShardsNotDone;
 
+		// NOTE: Don't finish here even if countShardsNotDone == 0. We need to dispatch tasks first.
+		// The completion check after dispatch (with dispatchedInThisIteration guard) prevents
+		// finishing in the same iteration we dispatch the last tasks.
 		if (countShardsNotDone == 0) {
-			TraceEvent("FileBackupSnapshotDispatchFinished")
+			TraceEvent("FileBackupSnapshotDispatchAllDoneBeforeDispatch")
 			    .detail("BackupUID", config.getUid())
-			    .detail("AllShards", countAllShards)
-			    .detail("ShardsDone", countShardsDone)
-			    .detail("ShardsNotDone", countShardsNotDone)
-			    .detail("SnapshotBeginVersion", snapshotBeginVersion)
-			    .detail("SnapshotTargetEndVersion", snapshotTargetEndVersion)
-			    .detail("CurrentVersion", recentReadVersion)
-			    .detail("SnapshotIntervalSeconds", snapshotIntervalSeconds);
-			Params.snapshotFinished().set(task, true);
```

Comment:
Where are we setting in the new code?

Reply:
3065       Params.snapshotFinished().set(task, true);

The old code had a bug - it would set snapshotFinished = true and return immediately when countShardsNotDone == 0 was detected before the dispatch loop ran. This could mark the snapshot as finished in the same iteration that dispatched the last batch of tasks, before those tasks actually completed; we'd were skipping out on doing the work.

The fix moves the completion check to after the dispatch loop and adds the !dispatchedInThisIteration guard. This ensures we only mark the snapshot as finished when:

 * All shards are done (countShardsNotDone == 0), AND
 * We didn't dispatch any new tasks in this iteration
 
This prevents race condition where could mark snapshot complete before last batch of tasks finish executing.